frontend/src/metabase/search/components/InfoText/use-info-text.tsx

```diff
@@ -0,0 +1,160 @@
+import { useMemo } from "react";
+import { t } from "ttag";
+import {
+  PLUGIN_COLLECTION_COMPONENTS,
+  PLUGIN_COLLECTIONS,
+} from "metabase/plugins";
+import type { WrappedResult } from "metabase/search/types";
+import type { Collection } from "metabase-types/api";
+import { useDatabaseQuery, useTableQuery } from "metabase/common/hooks";
+import {
+  browseDatabase,
+  browseSchema,
+  collection as collectionUrl,
+  tableRowsQuery,
+} from "metabase/lib/urls";
+
+import { Box } from "metabase/ui";
+import type TableType from "metabase-lib/metadata/Table";
+
+const { CollectionAuthorityLevelIcon } = PLUGIN_COLLECTION_COMPONENTS;
+
+export type InfoTextData = {
+  link?: string | null;
+  icon?: JSX.Element | null;
+  label: string | null;
+};
+
+export const useInfoText = (result: WrappedResult): InfoTextData[] => {
+  let infoTextHook;
+
+  switch (result.model) {
+    case "table":
+      infoTextHook = useTablePath;
+      break;
+    case "segment":
+    case "metric":
+      infoTextHook = useTableLink;
+      break;
+    case "collection":
+      infoTextHook = useCollectionInfoText;
+      break;
+    case "database":
+      infoTextHook = useDatabaseInfoText;
+      break;
+    case "action":
+      infoTextHook = useActionInfoText;
+      break;
+    case "card":
+    case "dataset":
+    case "indexed-entity":
+    default:
+      infoTextHook = useCollectionResultLink;
+      break;
+  }
+
+  return infoTextHook(result);
```

Comment:
:x: [Hooks should never be used conditionally.](https://react.dev/warnings/invalid-hook-call-warning#breaking-rules-of-hooks)

Reply:
>Don’t call Hooks inside loops, conditions, or nested functions. Instead, always use Hooks at the top level of your React function, before any early returns. You can only call Hooks while React is rendering a function component:
> ✅ Call them at the top level in the body of a [function component](https://react.dev/learn/your-first-component).
> ✅ Call them at the top level in the body of a [custom Hook](https://react.dev/learn/reusing-logic-with-custom-hooks).

I saw this coming 😄 Hooks shouldn't be _called_ conditionally, but we can call them at the top level of a custom hook. I'll take a look and see if there's any improvements to be made but this doesn't break the rules.

Reply:
> this doesn't break the rules.

The docs seem to have a loophole :) [Older docs provide a reason](https://legacy.reactjs.org/docs/hooks-rules.html):

> By following this rule, you ensure that **Hooks are called in the same order each time a component renders**. 

This is not ensured here.

Reply:
The refactoring is not that complex.
We need to lift the `switch` statement 1 level above `InfoTextAssetLink` and introduce 2 extra components that would use `useTableLink` and `useTablePath` respectively - those components would be rendered conditionally instead of `InfoTextAssetLink`. That's the gist of it I think.